x-pack/plugins/lens/public/xy_visualization/xy_expression.tsx

```diff
@@ -227,10 +241,27 @@ export function XYChart({
   histogramBarTarget,
   onClickValue,
   onSelectRange,
+  onControls,
 }: XYChartRenderProps) {
   const { legend, layers, fittingFunction, gridlinesVisibilitySettings } = args;
   const chartTheme = chartsThemeService.useChartsTheme();
   const chartBaseTheme = chartsThemeService.useChartsBaseTheme();
+  const chartRef = React.useRef<null | Chart>(null);
+
+  React.useEffect(() => {
+    if (!onControls) return;
+    const controls: LensChartControls = {
+      toPngBlob: async () => {
+        const snapshot = chartRef.current!.getPNGSnapshot({
+          backgroundColor: 'white',
+          pixelRatio: 2,
+        });
+        if (!snapshot) throw new Error('Could not generate PNG.');
+        return toBlob(snapshot.blobOrDataUrl);
+      },
+    };
+    onControls(controls);
```

Comment:
just call `controls.toPngBlob()` here ?

Reply:
You can, but that would generate a PNG of every chart on the dashboard at the moment they initially mount, not something you want.

Reply:
ehh ? onControls calls handlers.event() which in turn generates the png ...  so instead of doing handlers.event generate the png.

Reply:
in my code .png is generated only when user clicks "Copy PNG" action in UI

> onControls calls handlers.event() which in turn generates the png ...

`hadlers.event()` does not generate anything, it just stores the controls on embeddable, and then embeddable can use those controls when it needs to, i.e. when user clicks "Copy PNG" in UI.